[MM-18961] migrate help components to typescript

roseline124 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

I migrated help componenets to typescript. (in mattermost/mattermost-webapp)

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

Fixes https://github.com/mattermost/mattermost-server/issues/12466
- JIRA: https://mattermost.atlassian.net/browse/MM-18961

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->

Nothing I think.

roseline124 commented:
I'll fix lint!

devinbinnie commented:
Looks like you have one last lint error to fix! Almost there :)
```
/root/mattermost-webapp/components/help/help_controller.tsx
  6:1  error  There should be at least one empty line between import groups  import/order
```

lindy65 commented:
Thanks @roseline124 - is this something that can be covered by unit tests?

devinbinnie commented:
@lindy65 We should be okay for tests here, this is just a language migration.

@roseline124 Seems like we still have some lint errors:
```
/root/mattermost-webapp/components/help/help_controller.tsx
  5:1  error  There should be at least one empty line between import groups  import/order
```

FYI, you can run `make check-style` on your local machine to ensure that the code passes `eslint` before you push your code.

hmhealey commented on components/help/help_controller.tsx:
```diff
@@ -13,12 +11,17 @@ import Formatting from './components/formatting';
-    render() {
+export default class HelpController extends React.Component<HelpControllerProps, {}> {
+    public componentDidUpdate() {
+        (ReactDOM.findDOMNode(this) as HTMLInputElement).scrollIntoView();
```
I think this type might be incorrect. I'm pretty sure this will resolve to an `HTMLDivElement`. Also, would you mind switching this to use the parameterized version of `findDOMNode` instead of using `as`?

```suggestion
        ReactDOM.findDOMNode<HTMLDivElement>(this).scrollIntoView();
```

roseline124 commented:
if there are any problems, please tell me :)
thanks for helping me!

devinbinnie commented:
@roseline124 Linter is still failing:
```
/root/mattermost-webapp/components/help/help_controller.tsx
  22:48  error  Missing semicolon  semi
```

roseline124 commented:
> @roseline124 Linter is still failing:
> 
> ```
> /root/mattermost-webapp/components/help/help_controller.tsx
>   22:48  error  Missing semicolon  semi
> ```



> @roseline124 Linter is still failing:
> 
> ```
> /root/mattermost-webapp/components/help/help_controller.tsx
>   22:48  error  Missing semicolon  semi
> ```

fixed lint, thanks!

devinbinnie commented:
@roseline124 You have some tests failing:

```
Summary of all failing tests
 FAIL  utils/text_formatting_search_highlighting.test.jsx
  ● TextFormatting.searchHighlighting › search term highlighting in a code block

    expect(received).toEqual(expected) // deep equality

    Expected: "<div class=\"post-code post-code--wrap\"><code class=\"hljs\"><div class=\"post-code__search-highlighting\"><span class=\"search-highlight\">words</span> in a sentence</div>words in a sentence</code></div>"
    Received: "<div class=\"post-code post-code--wrap\"><code class=\"hljs\"><div class=\"post-code__search-highlighting\"><span class=\"search-highlight\">words</span> in a sentence
    </div>words in a sentence
    </code></div>"

      120 |             const output = TextFormatting.formatText(testCase.input, options).trim();
      121 | 
    > 122 |             expect(output).toEqual(testCase.expected);
          |                            ^
      123 |         });
      124 |     }
      125 | 

      at Object.toEqual (utils/text_formatting_search_highlighting.test.jsx:122:28)

  ● TextFormatting.searchHighlighting › search match highlighting in a code block

    expect(received).toEqual(expected) // deep equality

    Expected: "<div class=\"post-code post-code--wrap\"><code class=\"hljs\"><div class=\"post-code__search-highlighting\"><span class=\"search-highlight\">words</span> in a sentence</div>words in a sentence</code></div>"
    Received: "<div class=\"post-code post-code--wrap\"><code class=\"hljs\"><div class=\"post-code__search-highlighting\"><span class=\"search-highlight\">words</span> in a sentence
    </div>words in a sentence
    </code></div>"

      120 |             const output = TextFormatting.formatText(testCase.input, options).trim();
      121 | 
    > 122 |             expect(output).toEqual(testCase.expected);
          |                            ^
      123 |         });
      124 |     }
      125 | 

      at Object.toEqual (utils/text_formatting_search_highlighting.test.jsx:122:28)

 FAIL  utils/message_html_to_component.test.jsx
  ● messageHtmlToComponent › latex

    expect(received).toMatchSnapshot()

    Snapshot name: `messageHtmlToComponent latex 1`

    - Snapshot
    + Received

    @@ -3,14 +3,16 @@
          This is some latex!
        </p>,
        "
      ",
        <LatexBlock
    -     content="x^2 + y^2 = z^2"
    +     content="x^2 + y^2 = z^2
    + "
        />,
        <LatexBlock
    -     content="F_m - 2 = F_0 F_1 \\dots F_{m-1}"
    +     content="F_m - 2 = F_0 F_1 \\dots F_{m-1}
    + "
        />,
        <p>
          That was some latex!
        </p>,
      ]

      26 |         const html = TextFormatting.formatText(input);
      27 | 
    > 28 |         expect(messageHtmlToComponent(html)).toMatchSnapshot();
         |                                              ^
      29 |     });
      30 | 
      31 |     test('link without enabled tooltip plugins', () => {

      at Object.toMatchSnapshot (utils/message_html_to_component.test.jsx:28:46)


Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or run `npm run test-ci -- -u` to update them.
```

Please make sure to run `make check-types`, `make check-style` and `make test` before you push your code up to ensure it passes all the checks.

hmhealey commented:
It looks like those failing tests are related to the Markdown changes I made recently. Your code changes don't appear to have affected that area, but it looks like you might've accidentally reverted those changes in the `package-lock.json` on your branch.

Could you try reverting `package-lock.json` to match the version on master? If you update your `master` branch to match the upstream one, you can run `git checkout master -- package-lock.json` to overwrite that.

Let us know if you have any questions or could use some help doing that. `git` and npm's lock files can definitely be tricky to work with.

roseline124 commented:
> It looks like those failing tests are related to the Markdown changes I made recently. Your code changes don't appear to have affected that area, but it looks like you might've accidentally reverted those changes in the `package-lock.json` on your branch.
> 
> Could you try reverting `package-lock.json` to match the version on master? If you update your `master` branch to match the upstream one, you can run `git checkout master -- package-lock.json` to overwrite that.
> 
> Let us know if you have any questions or could use some help doing that. `git` and npm's lock files can definitely be tricky to work with.

Thank you so much! 
I reverted package-lock.json with package-lock.json of origin/master branch (the file is same with before I changed).
Is that right? Or I need to update with package-lock.json of upstream/master branch?

devinbinnie commented:
@roseline124 You need to update it to match `upstream/master`.

mattermod commented:
This issue has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @hanzei

hanzei commented:
Hey @roseline124, can we help you with merging `master`?

hanzei commented:
@roseline124 You have some type errors in your code. Let me know if you need help resolving them.

Please note that we will close this PR after another 10 days of inactivity. If you don't object, other contributors might complete this PR for you.

M-ZubairAhmed commented:
Please assign it to me after the due waiting period (10 days) gets completed

devinbinnie commented:
@roseline124 Is it alright for @M-ZubairAhmed to finish off this PR for you?

/cc @hanzei

hanzei commented:
I'm going to close this PR due to inactivity. 

@M-ZubairAhmed It would be great if you could take over the HW.

M-ZubairAhmed commented:
Yes sure i would love to
